dali/kernels/imgproc/convolution/filter_gpu.cuh

```diff
@@ -0,0 +1,634 @@
+  int h, wc;
+  int64_t hwc;
+};
+
+/**
+ * @brief Creates ROI description for input and output of the same shape.
+ */
+struct ROIFull {
+  DALI_DEVICE DALI_FORCEINLINE ROI operator()(const ShapeDesc& shape_desc) {
+    return {0, shape_desc.h, 0, shape_desc.wc, shape_desc.h, shape_desc.wc, shape_desc.hwc};
+  }
+};
+
+/**
+ * @brief Creates (maximal) ROI shrunk so that all filter positions lie fully within the image.
+ */
+struct ROIOnlyValid {
+  DALI_DEVICE DALI_FORCEINLINE ROI operator()(const ShapeDesc& shape_desc) {
+    int h_begin = -shape_desc.filter_top_anchor;
+    int h_end = shape_desc.h - (shape_desc.r - 1 + shape_desc.filter_top_anchor);
+    int wc_begin = (-shape_desc.filter_left_anchor) * shape_desc.c;
+    int wc_end = shape_desc.wc - (shape_desc.s - 1 + shape_desc.filter_left_anchor) * shape_desc.c;
+    int h = h_end - h_begin;
+    int wc = wc_end - wc_begin;
+    return {h_begin, h_end, wc_begin, wc_end, h, wc, h * wc};
+  }
+};
+
+/** @} */  // end of ConvolutionROI
+
+
+/** @defgroup InputLoader InputLoader is meant to specialize loading of the sample from global
+ * memory. This is how different border modes (apart from BORDER_VALID) are handled.
+ * @{
+ */
+
+template <bool degenerated_extents>
+struct Reflect101 {
+  DALI_HOST_DEV DALI_FORCEINLINE int border_remap(int idx, int len) const {
+    assert(len > 0);
+    return boundary::idx_reflect_101(idx, len);
+  }
+};
+
+struct Reflect1001 {
+  DALI_HOST_DEV DALI_FORCEINLINE int border_remap(int idx, int len) const {
+    assert(len > 0);
+    return boundary::idx_reflect_1001(idx, len);
+  }
+};
+
+struct Clamp {
+  DALI_HOST_DEV DALI_FORCEINLINE int border_remap(int idx, int len) const {
+    assert(len > 0);
+    return boundary::idx_clamp(idx, 0, len);
+  }
+};
+
+struct Wrap {
+  DALI_HOST_DEV DALI_FORCEINLINE int border_remap(int idx, int len) const {
+    assert(len > 0);
+    return boundary::idx_wrap(idx, len);
+  }
+};
+
+template <typename Remap, typename In>
+struct InLoaderBorderRemap : protected Remap {
+  DALI_HOST_DEV DALI_FORCEINLINE int remap_height(int idx, const ShapeDesc& sample_shape) const {
+    return this->border_remap(idx, sample_shape.h);
+  }
+
+  DALI_HOST_DEV DALI_FORCEINLINE int remap_width(int idx, const ShapeDesc& sample_shape) const {
+    return border_remap_strided(idx, sample_shape.w, sample_shape.c, sample_shape.wc);
+  }
+
+  DALI_HOST_DEV DALI_FORCEINLINE In load(const In* __restrict__ in, int y, int x,
+                                         const ShapeDesc& sample_shape) const {
+    return in[y * static_cast<int64_t>(sample_shape.wc) + x];
+  }
+
+ protected:
+  DALI_HOST_DEV DALI_FORCEINLINE int border_remap_strided(int idx, int reflect_dim_size,
+                                                          int inner_stride,
+                                                          int total_stride) const {
+    if (idx < 0) {
+      int reflect_dim_idx = (idx + 1) / inner_stride - 1;
```

Comment:
I'd add some simple explanation here for future readers, perhaps with some example:
```
// First, shift by 1 towards 0, so that indices belonging to the same pixel translate to the right pixel index when dividing by the stride -  (-6, -5, -4, -3, -2, -1) + 1 -> (-5, -4, -3. -2, -1, 0)
// (-5, -4, -3, -2, -1, 0) / 3 -> (1, 1, 1, 0, 0, 0)
// Now, shift away from 0 to revert (1, 1, 1, 0, 0, 0) + 1 -> (2, 2, 2, 1, 1, 1)
// We don't want to reflect the order of the channels, so we revert the order
// (-2, -1, 0, -2, -1, 0) + 3 - 1 -> (0, 1, 2, 0, 1, 2)
```